tests/components/upb/test_config_flow.py

```diff
@@ -0,0 +1,152 @@
+"""Test the UPB Control config flow."""
+
+from asynctest import MagicMock, PropertyMock, patch
+
+from homeassistant import config_entries, setup
+from homeassistant.components.upb.const import DOMAIN
+
+
+def mocked_upb(sync_complete=True, config_ok=True):
+    """Mock UPB lib."""
+
+    def _upb_lib_connect(callback):
+        callback()
+
+    upb_mock = MagicMock()
+    type(upb_mock).network_id = PropertyMock(return_value="42")
+    type(upb_mock).config_ok = PropertyMock(return_value=config_ok)
+    if sync_complete:
+        upb_mock.connect.side_effect = _upb_lib_connect
+    return patch(
+        "homeassistant.components.upb.config_flow.upb_lib.UpbPim", return_value=upb_mock
+    )
+
+
+async def valid_tcp_flow(hass, sync_complete=True, config_ok=True):
+    """Get result dict that are standard for most tests."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    with mocked_upb(sync_complete, config_ok), patch(
+        "homeassistant.components.upb.async_setup_entry", return_value=True
+    ):
+        flow = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+        result = await hass.config_entries.flow.async_configure(
+            flow["flow_id"],
+            {"protocol": "TCP", "address": "1.2.3.4", "file_path": "upb.upe"},
+        )
+    return result
+
+
+async def test_full_upb_flow_with_serial_port(hass):
+    """Test a full UPB config flow with serial port."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+
+    with mocked_upb(), patch(
+        "homeassistant.components.upb.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "homeassistant.components.upb.async_setup_entry", return_value=True
+    ) as mock_setup_entry:
+        flow = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+
+        result = await hass.config_entries.flow.async_configure(
+            flow["flow_id"],
+            {
+                "protocol": "Serial port",
+                "address": "/dev/ttyS0:115200",
+                "file_path": "upb.upe",
+            },
+        )
+
+    assert flow["type"] == "form"
+    assert flow["errors"] == {}
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)